Task Liverpool/src/main/java/pages/BasePage.java

```diff
@@ -0,0 +1,53 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.time.Duration;
+import java.util.List;
+
+public class BasePage {
+
+    private WebDriver webDriver;
+    private WebDriverWait wait;
+
+    protected BasePage(WebDriver webDriver){
+        this.webDriver = webDriver;
+        wait = new WebDriverWait(webDriver, Duration.ofSeconds(10,1));
+    }
+
+    protected WebDriver getWebDriver(){
+        return webDriver;
+    }
+
+    protected WebElement findElement(By locator){
+        wait.until(ExpectedConditions.presenceOfElementLocated(locator));
+        return getWebDriver().findElement(locator);
+    }
+
+    protected List<WebElement> findElements(By locator){ ///FIX THIS
+        wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(locator));
+        return getWebDriver().findElements(locator);
+    }
+
+    protected void click(By locator){
+        WebElement element = wait.until(ExpectedConditions.elementToBeClickable(locator));
+        element.click();
+    }
+
+    protected void type(By locator, String text){
+        findElement(locator).sendKeys(text);
+    }
+
+    protected String getElementText(By locator){
+        WebElement element = wait.until(ExpectedConditions.presenceOfElementLocated(locator));
+        return element.getText();
+    }
+
+    protected String getElementAttributeOfAList(WebElement productImageElement, String attribute){
```

Comment:
You are using a specific attribute name (productImageElement) in a generic method. Also (and this will depend of the context) I think that it will be better to have one method to get the value of the element attribute instead of doing it for an entire list.

Reply:
agree with Ramses, you are getting from one element not of a List, also I would like to check first if the element has the attribute to retrieve (Nos sure what happens if `getAttribute` does not find it)

Reply:
Correct, the attribute name should not be specific in that case, I changed it to "element". Also changed the method name to "getElementAttribute" (removed "ofAList").
Currently if the attribute is not found, it returns "null" and that is saved in my urlList. Which is Not correct because that way the test still passes. I am looking for a way to handle this.